src/services/pool-creator/pool-creator.service.ts

```diff
@@ -0,0 +1,167 @@
+import { toNormalizedWeights } from '@balancer-labs/balancer-js';
+
+type Address = string;
+
+export interface CreatePoolReturn {
+  id: string;
+  address: Address;
+}
+
+const JOIN_KIND_INIT = 0;
+
+export interface JoinPoolRequest {
+  assets: Address[];
+  maxAmountsIn: string[];
+  userData: any;
+  fromInternalBalance: boolean;
+}
+
+export class PoolCreator {
+  public async createWeightedPool(
+    provider: Web3Provider,
+    name: string,
+    symbol: string,
+    swapFee: string,
+    tokens: TokenWeight[],
+    owner: Address
+  ): Promise<CreatePoolReturn> {
+    const weightedPoolFactoryAddress =
+      configService.network.addresses.weightedPoolFactory;
+
+    // Tokens must be sorted in order of addresses
+    tokens = this.sortTokens(tokens);
+
+    const tokenAddresses: Address[] = tokens.map((token: TokenWeight) => {
+      return token.tokenAddress;
+    });
+
+    const tokenWeights = this.calculateTokenWeights(tokens);
+    const swapFeeScaled = new BigNumber(`${swapFee}e16`);
+
+    const params = [
+      name,
+      symbol,
+      tokenAddresses,
+      tokenWeights,
+      swapFeeScaled.toString(),
+      owner
+    ];
+
+    const tx: TransactionResponse = await sendTransaction(
+      provider,
+      weightedPoolFactoryAddress,
+      WeightedPoolFactory__factory.abi,
+      'create',
+      params
+    );
+
+    const receipt: any = await tx.wait();
+    const events = receipt.events.filter(e => e.event === 'PoolCreated');
+    const poolAddress = events[0].args[0];
+
+    const pool = new Contract(poolAddress, WeightedPool__factory.abi, provider);
+    const poolId = await pool.getPoolId();
+
+    const poolDetails: CreatePoolReturn = {
+      id: poolId,
+      address: poolAddress
+    };
+
+    return poolDetails;
+  }
+
+  public async joinPool(
+    provider: Web3Provider,
+    poolId: string,
+    sender: Address,
+    receiver: Address,
+    tokens: TokenWeight[],
+    initialBalances: BigNumber[]
+  ): Promise<TransactionReceipt> {
+    const initialBalancesString: string[] = initialBalances.map(n =>
+      n.toString()
+    );
+    const initUserData = defaultAbiCoder.encode(
+      ['uint256', 'uint256[]'],
+      [JOIN_KIND_INIT, initialBalancesString]
+    );
+
+    const tokenAddresses: Address[] = tokens.map((token: TokenWeight) => {
+      return token.tokenAddress;
+    });
+
+    const joinPoolRequest: JoinPoolRequest = {
+      assets: tokenAddresses,
+      maxAmountsIn: initialBalancesString,
+      userData: initUserData,
+      fromInternalBalance: false
+    };
+
+    // Get approval for all tokens here
+
+    const vaultAddress = configService.network.addresses.vault;
+    const tx = await sendTransaction(
+      provider,
+      vaultAddress,
+      Vault__factory.abi,
+      'joinPool',
+      [poolId, sender, receiver, joinPoolRequest]
+    );
+
+    const receipt: TransactionReceipt = await tx.wait();
+
+    return receipt;
```

Comment:
This service should return the transaction promise and the frontend should handle the transaction so we can add it to the notifications etc.
```suggestion
    return await sendTransaction(
      provider,
      vaultAddress,
      Vault__factory.abi,
      'joinPool',
      [poolId, sender, receiver, joinPoolRequest]
    );
```

Reply:
What about for the create? Issue with create is it's two steps to first create the pool then retrieve its ID.

Reply:
I've answered below, if there are two steps, e.g. create and fetch id, then there should really be two functions, and the frontend should be responsible for waiting for the first and then calling the second.

Reply:
That makes sense